### PR TITLE
[CNSMR-2107] Create UnitTestingViewModels.md

### PR DESCRIPTION
**Pull request by ilyapuchka**

### CODE REVIEW

**Comment by karabatov** [comment]
I wanted to write that :triggered:

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.
+
+- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
+After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release" the first event and so on. 
+It is also possible to advance scheduler by specific time interval with `advance(by:)` method instead of using `sleep`. This is useful when any signal in the view model state machine uses timer (i.e. with `delay` operator)
+
+Example:
+
```
Comment (dmcrodrigues) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]: Maybe we could have an example of `stub` too?

**Review by dmcrodrigues — APPROVED** [review]
Good stuff @ilyapuchka 👏

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 1 above)
Comment (ilyapuchka) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: essentially it should be just calling `makeViewMode` with some parameters, and that is already in the template, that's why I didn't put it there 🤷‍♂

Comment (dmcrodrigues) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: 👍

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+- routing events
+- analytics events
+
+### Creating a new test case
+
+To create a test case for a view model use an Xcode template called `BentoViewModelTests` and provide the view model name (without `ViewModel` suffix). This template will create a basic structure for the test case that you will need to adjust to your specific implementation.
+
+### Writing a test
+
+When implementing a test use `perform(stub:when:assert:)` herlper method from `BabylonTestUtilities`.
```
Comment (karabatov) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
When implementing a test use `perform(stub:when:assert:)` helper method from `BabylonTestUtilities`.
```

Comment (karabatov) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: Maybe you could provide the full signature here? Because not having it memorised it's quite hard to follow (and this is aimed at new starters mostly, right?).

I think the reader should be able to have a more or less complete understanding of what is happening without having to look at the code in the project, and this doc only has truncated signatures and a non-structured description of parameters. I believe providing a complete, if short, example first, with full signatures so that the reader understands what's going on, and then dissecting it in following sections would be more conductive to learning.

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.
```
Comment (karabatov) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events dispatch via its `advance` method later in the `when` closure.
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model.
```
Comment (karabatov) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
```

**Review by karabatov — APPROVED** [review]

**Comment by karabatov** [comment]
Can you please put the ticket number in the description if this is towards the H2 ticket? I'm trying to understand how WIP it is and what else will be covered to avoid duplicating effort on the backend feature doc :)

**Review by viorel15 — APPROVED** [review]

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
+## Unit testing view models
+
+### What do we test in the view model unit tests?
+
+- view model state changes
+- routing events
+- analytics events
```
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]: Should they be in a different file/test? IMO it makes it easier to read/scan.

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Creating a new test case
+
+To create a test case for a view model use an Xcode template called `BentoViewModelTests` and provide the view model name (without `ViewModel` suffix). This template will create a basic structure for the test case that you will need to adjust to your specific implementation.
```
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
To create a test case for a view model use the Xcode template `BentoViewModelTests` and provide the view model name (without `ViewModel` suffix). This template will create a basic structure for the test case that you will need to adjust to your specific implementation.

E.g: your feature is called `MyFeature`, and is has a `MyFeatureViewModel`. You just need to name the file `MyFeature`, the template will automatically add the proper suffix making it `MyFeatureViewModelTests`.
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 2 above)
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
When implementing a test use the `perform(stub:when:assert:)` helper method from `BabylonTestUtilities`.
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 3 above)
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- the `stub` parameter is used to create a view model and to provide it its stubbed dependencies via the `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, which allows to control signals events dispatch via its `.advance` method (to be used in later moment in the `when` closure).
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 4 above)
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- the `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `.send` method and provide it with an action. This effectively simulates user interaction with a screen managed by this view model. 
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+### Writing a test
+
+When implementing a test use `perform(stub:when:assert:)` herlper method from `BabylonTestUtilities`.
+
+- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.
+
+- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
+After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release" the first event and so on.
```
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release"/trigger the first event and so on. 
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.
+
+- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
+After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release" the first event and so on. 
+It is also possible to advance scheduler by specific time interval with `advance(by:)` method instead of using `sleep`. This is useful when any signal in the view model state machine uses timer (i.e. with `delay` operator)
```
Comment (diegopetrucci) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
It is also possible to advance the scheduler by a specific time interval with `advance(by:)` method instead of using `sleep`. This is useful when any signal in the view model state machine uses a timer (i.e. with `delay` operator)
```

**Review by diegopetrucci — APPROVED** [review]
Feel free to tweak my suggestions in case you decide to use them

**Review by horothesun — APPROVED** [review]

**Review by emeseuk — APPROVED** [review]

**Review by mluisbrown — APPROVED** [review]

**Review by adiki — APPROVED** [review]

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 6 above)
Comment (daniloaliberti) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: @diegopetrucci 
> ... and is has ... 

... and *it* has ... would you mind if I edit your suggestion?

Comment (daniloaliberti) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: also before the "and" no comma is needed.

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 4 above)
Comment (daniloaliberti) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]: All other sections have code example. Could we have here as well? It's very helpful. I just fully understand seeing an example. 😬

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 6 above)
Comment (diegopetrucci) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: Sure big man, feel free

Comment (daniloaliberti) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: just to be clear: I'll do it just when it's a typo in a suggestion. Otherwise it's a rabbit hole. 😉

**Review by daniloaliberti — APPROVED** [review]
I'm happy with all here. 
Wanted a code example in the `stub` part. (up-to-you)
And all the typos were found before, so no point making the 3rd batch of suggestions here. 

Would it be possible to include the UnitTest template in the Bento template?
In the `Tests` folder. (out-of-PR-scope)
🤨

Overall it's a very good doc! ✨

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 2 above)
Comment (ilyapuchka) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: do you thin `perform<ViewModel: BoxViewModel>(stub: (TestScheduler) -> ViewModel, when: (ViewModel, TestScheduler) -> Void, assert: ([ViewModel.State]) -> Void)` will make it simpler to understand? I have doubts 😄

File: Cookbook/Technical-Documents/UnitTestingViewModels.md (hunk 5 above)
Comment (ilyapuchka) [reply Cookbook/Technical-Documents/UnitTestingViewModels.md]: this can lead to unnecessary duplicated tests, as the same action can result in multiple events (state change, routing, analytics event) happening all together

**Comment by ilyapuchka** [comment]
> Would it be possible to include the UnitTest template in the Bento template?
In the Tests folder. (out-of-PR-scope)

I have it in mind, it will be a good addition @daniloaliberti 

@karabatov I've updated the example and now it is one solid method call, should be more clear now

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        ]))
+    
+        expect(analytics.accumulatedEvents).to(equal([
+            AnyEquatable(Tracking.ActionEvents.event),
+        ]))
+    }
+)
+```
+
+- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.
```
Comment (karabatov) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via its `advance` method later in the `when` closure.
```

File: Cookbook/Technical-Documents/UnitTestingViewModels.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model.
```
Comment (karabatov) [inline Cookbook/Technical-Documents/UnitTestingViewModels.md]:
```suggestion
- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
```

**Review by karabatov — APPROVED** [review]
You'll also want to add a link to the Cookbook ;)

**Review by jtomanik — APPROVED** [review]
